common/hashing.h

```diff
@@ -0,0 +1,721 @@
+class Hasher {
+ public:
+  Hasher(Hasher&& arg) = default;
+  Hasher(const Hasher& arg) = delete;
+  auto operator=(Hasher&& rhs) -> Hasher& = default;
+
+  // Extracts the current state as a `HashCode` for use.
+  explicit operator HashCode() const { return HashCode(buffer); }
+
+  // Convenience method to hash an object by its object representation when that
+  // is known to be valid. This is primarily useful for builtin and primitive
+  // types.
+  //
+  // This can be directly used for simple users combining some aggregation of
+  // objects. However, when possible, prefer the tuple version below for
+  // aggregating several primitive types into a hash.
+  template <typename T, typename = std::enable_if_t<
+                            std::has_unique_object_representations_v<T>>>
+  static auto Hash(Hasher hash, const T& value) -> Hasher;
+
+  // Convenience method to hash a variable number of objects whose object
+  // representation when that is known to be valid. This is primarily useful for
+  // for builtin and primitive type objects.
+  //
+  // There is no guaranteed correspondence between the behavior of a single call
+  // with multiple parameters and multiple calls. This routine is also optimized
+  // for handling relatively small numbers of objects. For hashing large
+  // aggregations, consider some Merkle-tree decomposition or arranging for a
+  // byte buffer that can be hashed as a single buffer.
+  template <typename... Ts,
+            typename = std::enable_if_t<
+                (... && std::has_unique_object_representations_v<Ts>)>>
+  static auto Hash(Hasher hash, const Ts&... value) -> Hasher;
+
+  // The simplest three more primitive APIs to use when incorporating state. The
+  // first two incorporate one or two unsized components of state represented as
+  // a collection of 64 bits.
+  static auto HashOne(Hasher hash, uint64_t data) -> Hasher;
+  static auto HashTwo(Hasher hash, uint64_t data0, uint64_t data1) -> Hasher;
+
+  // This is a heavily optimized routine for incorporating a dynamically sized
+  // sequence of bytes into the state. It has carefully structured inline code
+  // paths for short byte sequences and a reasonably high bandwidth code path
+  // for longer sequences. The size of the byte sequence is always incorporated
+  // into the hash.
+  static auto HashSizedBytes(Hasher hash, llvm::ArrayRef<std::byte> bytes)
+      -> Hasher;
+
+  // Even lower-level primitive APIs to facilitate reading specific sized data
+  // efficiently into one or two 64-bit values
+  static auto Read1(const std::byte* data) -> uint64_t;
+  static auto Read2(const std::byte* data) -> uint64_t;
+  static auto Read4(const std::byte* data) -> uint64_t;
+  static auto Read8(const std::byte* data) -> uint64_t;
+  static auto Read1To3(const std::byte* data, ssize_t size) -> uint64_t;
+  static auto Read4To8(const std::byte* data, ssize_t size) -> uint64_t;
+  static auto Read8To16(const std::byte* data, ssize_t size)
```

Comment:
Looks like in the code that when size == 8 it always uses `Read4To8` instead. I think it is helpful to readers to name these to clarify which of the two functions are used in that case:
```suggestion
  static auto Read9To16(const std::byte* data, ssize_t size)
```

Reply:
Hmm, not sure.

This either function is *valid* to call for size 8 -- it's an optimization question which one we use. And there might be reasons to switch.

Basically, the functions are named after their contract not the particular use cases. WDYT?

Reply:
Currently it looks like this function is never used for size 8, and it looks like it would be a performance regression to do so. I guess this depends on your reading of the word "efficiently" in the comment. It seems like if you want to say the name reflects the legal values of size, then leave the name, otherwise it seems like it is misleading since that is not how these functions are actually used.

Reply:
Based on discussion it seems like what I'm looking for is an update to the comments that make it clear that these are named based on the requirements of the functions, and they are specifically for cases where the size is unknown, but known to be in a range.

Reply:
Hopefully the useful comment is now added.